modules/publirBidAdapter.js

```diff
@@ -0,0 +1,389 @@
+const DEFAULT_SELLER_ENDPOINT = 'https://prebid.publir.com/publirPrebidEndPoint';
+const DEFAULT_IMPS_ENDPOINT = 'https://prebidimpst.publir.com/publirPrebidImpressionTracker';
+const SUPPORTED_SYNC_METHODS = {
+  IFRAME: 'iframe',
+  PIXEL: 'pixel'
+}
+
+export const storage = getStorageManager({ bidderCode: BIDDER_CODE });
+export const spec = {
+  code: BIDDER_CODE,
+  version: ADAPTER_VERSION,
+  aliases: ['plr'],
+  supportedMediaTypes: SUPPORTED_AD_TYPES,
+  isBidRequestValid: function (bidRequest) {
+    if (!bidRequest.params.pubId) {
+      logWarn('pubId is a mandatory param for Publir adapter');
+      return false;
+    }
+
+    return true;
+  },
+  buildRequests: function (validBidRequests, bidderRequest) {
+    const reqObj = {};
+    const generalObject = validBidRequests[0];
+    reqObj.params = generatePubGeneralsParams(generalObject, bidderRequest);
+    reqObj.bids = generatePubBidParams(validBidRequests, bidderRequest);
+    reqObj.bids.timestamp = timestamp();
+    let options = {
+      withCredentials: false
+    };
+
+    return {
+      method: 'POST',
+      url: DEFAULT_SELLER_ENDPOINT,
+      data: reqObj,
+      options
+    }
+  },
+  interpretResponse: function ({ body }) {
+    const bidResponses = [];
+    if (body.bids) {
+      body.bids.forEach(adUnit => {
+        const bidResponse = {
+          requestId: adUnit.requestId,
+          cpm: adUnit.cpm,
+          currency: adUnit.currency || CURRENCY,
+          width: adUnit.width,
+          height: adUnit.height,
+          ttl: adUnit.ttl || TTL,
+          creativeId: adUnit.creativeId,
+          netRevenue: adUnit.netRevenue || true,
+          nurl: adUnit.nurl,
+          mediaType: adUnit.mediaType,
+          meta: {
+            mediaType: adUnit.mediaType
+          },
+        };
+
+        if (adUnit.mediaType === VIDEO) {
+          bidResponse.vastXml = adUnit.vastXml;
+        } else if (adUnit.mediaType === BANNER) {
+          bidResponse.ad = adUnit.ad;
+        }
+
+        if (adUnit.adomain && adUnit.adomain.length) {
+          bidResponse.meta.advertiserDomains = adUnit.adomain;
+        } else {
+          bidResponse.meta.advertiserDomains = [];
+        }
+        if (adUnit?.meta?.ad_key) {
+          bidResponse.meta.ad_key = adUnit.meta.ad_key ?? null;
+        }
+        if (adUnit.campId) {
+          bidResponse.campId = adUnit.campId;
+        }
+        bidResponse.bidder = BIDDER_CODE;
+        bidResponses.push(bidResponse);
+      });
+    } else {
+      return [];
+    }
+    return bidResponses;
+  },
+  getUserSyncs: function (syncOptions, serverResponses) {
+    const syncs = [];
+    for (const response of serverResponses) {
+      if (syncOptions.iframeEnabled && response.body.params.userSyncURL) {
+        syncs.push({
+          type: 'iframe',
+          url: response.body.params.userSyncURL
+        });
+      }
+      if (syncOptions.pixelEnabled && isArray(response.body.params.userSyncPixels)) {
```

Comment:
Last one I think… you don't valid body.params exists, it throws an error when your response is empty

![Capture d’écran 2024-03-06 à 12 02 54](https://github.com/prebid/Prebid.js/assets/16453882/06ced945-ec0b-4483-b729-6fbbfe406fb2)

Reply:
As per the feedback, modified the code.

Reply:
I apologize, but you also have to do it for userSyncURL